### ISSUE TITLE
Log4NetProvider is not shutting down repository

### ISSUE THREAD

**Issue by ydhondt** [huorswords/Microsoft.Extensions.Logging.Log4Net.AspNetCore#68]
When using the `Log4NetProvider` with an appender (e.g. ApplicationInsightsAppender) which buffers messages before persisting them, messages in that buffer are lost when the application exits.

This comes becomes the appenders are not correctly closed on exit. 

The problem is with the [Dispose()](https://github.com/huorswords/Microsoft.Extensions.Logging.Log4Net.AspNetCore/blob/develop/src/Microsoft.Extensions.Logging.Log4Net.AspNetCore/Log4NetProvider.cs#L103-L111) method where you do not clean up everything.

The repository you created is not shut down correctly. You really need to force the repository to shut down as it does not implement `IDisposable`. When doing so it will close the appenders (and flush their buffers). So, something along the following lines:

```csharp
        protected virtual void Dispose(bool disposing)
        {
            this.loggerRepository.Shutdown();
            this.loggers.Clear();
        }
```

Note that I removed the `disposing` check because it does actually the opposite of what it should do. That return statement would only make sense if the instance was previously disposed of.  Otherwise, you would still want to clean things up even if they are all managed resources (see [implementing dispose](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose)).

**Comment by huorswords** [comment]
Hello @ydhondt,

Thank you very much by reporting this bug. You are right, the dispose pattern is not correctly implemented and I'm not considering the logger repository shutdown on the dispose code.

I have fixed it on this https://github.com/huorswords/Microsoft.Extensions.Logging.Log4Net.AspNetCore/commit/82eabdbdbe578cc99fef7564fd925a245096ae6e commit.

The fixed version will be uploaded as version 2.2.12 into nuget.org during the day.

Best regards,
--
Angel

**Comment by huorswords** [comment]
Published! Closing the bug.
